### PR TITLE
[JENKINS-62332] Show when a plugin version newer than offered exists

### PR DESCRIPTION
**Pull request by daniel-beck**
See [JENKINS-62332](https://issues.jenkins-ci.org/browse/JENKINS-62332) and https://github.com/jenkins-infra/update-center2/pull/382

This adds notices to entries on the 'Updates' and 'Available' tabs of the plugin manager when a newer version exists than is being offered for installation.

If the instance has the latest offered plugin release installed, new virtual (uninstallable) entries are created on the 'Updates' tab to inform users that something newer exists. On one hand I like that this shows unavailable content right where it would appear, OTOH it could be weird or confusing, especially if admins don't read. Feedback welcome.

<details>
<summary>Screenshots</summary>

#### Before

Screenshot 2020-05-25 at 22 39 03

#### After

Screenshot 2020-05-25 at 22 38 40
</details>

### Testing notes

Replace the update site URL in _Plugin Manager » Advanced_ with `https://updates.jenkins.io/2.204/update-center.json`

### Proposed changelog entries

* Show in plugin manager when newer releases of plugins exist but aren't being offered due to unsatisfied requirements.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/resources/hudson/PluginManager/table.properties
Comment (res0nance) [inline core/src/main/resources/hudson/PluginManager/table.properties:56]:
```suggestion
  This is typically the case when plugin's requirements are not satisfied. e.g. a more recent version of Jenkins
```

File: core/src/main/resources/hudson/PluginManager/table.properties
Comment (res0nance) [inline core/src/main/resources/hudson/PluginManager/table.properties:59]:
```suggestion
  This is typically the case when plugin's requirements are not satisfied. e.g. a more recent version of Jenkins. \
```

**Review by res0nance — APPROVED** [review]
I think the idea is fine. It can help people who see there is a newer version on our plugins site but is not available in their plugin manager.

Included some suggestion for better readability

**Review by fqueiruga — COMMENTED** [review]
I find it a bit confusing to just show the entry. Many people will just wonder why they cannot install the plugin and will not bother to read a whole paragraph to understand. 

Maybe we could add a short message like `Not compatible` next to the plugin name?

File: core/src/main/java/hudson/PluginManager.java
Comment (varyvol) [inline core/src/main/java/hudson/PluginManager.java:2274]: Isn't this taken care of in the plugin.latest calculation?

**Review by varyvol — COMMENTED** [review]
I agree with @fqueiruga : users won't really understand the reasons for the plugin not to be available. They will ignore the message or complain because they think their instance is not working properly.

File: core/src/main/java/hudson/PluginManager.java
Comment (daniel-beck) [reply core/src/main/java/hudson/PluginManager.java:2274]: This code supports `latest` being added to JSON even if the latest release is being offered. While that's not the behavior of update-center2 in the linked PR, I wouldn't consider it wrong.

Comment (varyvol) [reply core/src/main/java/hudson/PluginManager.java:2274]: Maybe it's not wrong, but isn't it more sensible to make both sides have the same behaviour? I believe it's easier to undertand.

Comment (daniel-beck) [reply core/src/main/java/hudson/PluginManager.java:2274]: https://en.wikipedia.org/wiki/Robustness_principle

**Comment by daniel-beck** [comment]
New version:

> Screenshot

CSS is mostly copied from the normal label tags.

**Comment by fqueiruga** [comment]
I think it's much more clear this way 👍 . Although I think all caps it's a bit too much. 

Did you try to make some common CSS badge code and then create variants?

**Comment by daniel-beck** [comment]
@fqueiruga 

> Did you try to make some common CSS badge code and then create variants?

No. Didn't seem worth it until I get some preliminary/conditional approval for the basic approach.

**Review by fqueiruga — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]
I tested it locally, and it looks pretty good. Thanks @daniel-beck !

I plan to merge it in 24 hours if no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information

**Comment by daniel-beck** [comment]
> > Did you try to make some common CSS badge code and then create variants?
> 
> No. Didn't seem worth it until I get some preliminary/conditional approval for the basic approach.

Looked at it and am uncertain how this would work given the numerous different pseudo classes etc. involved. Will leave this for someone who knows what they're doing.

**Review by res0nance — APPROVED** [review]

**Comment by daniel-beck** [comment]
Update center change is live, so this can now be tested more easily: Replace the update site URL in _Plugin Manager » Advanced_ with `https://updates.jenkins.io/2.204/update-center.json`

**Comment by daniel-beck** [comment]
There's a bug in here I'd still like to fix: If only unavailable updates exist, the "Updates" tab shows nothing. Culprit: https://github.com/jenkinsci/jenkins/blob/fc4b0c36d85a1a9d01bf1cda8ec5e6c6e53f0db8/core/src/main/resources/hudson/PluginManager/table.jelly#L76

**Comment by daniel-beck** [comment]
Fixed.

All unavailable:

> Screenshot

Some unavailable:

> Screenshot

All available (the coloring is `already-upgraded` as I haven't restarted since installing):

> 
Screenshot

File: core/src/main/resources/hudson/PluginManager/_table.css
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/PluginManager/_table.css:98]:
```suggestion
}

```

**Review by oleg-nenashev — APPROVED** [review]
LGTM, thanks for catching and fixing the bug

**Review by fqueiruga — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
We may merge it in 24 hours if no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information

**Pull request merged**
